### PR TITLE
CAD-4886: Replace LazyTemplateProvider with tcomb interface

### PR DESCRIPTION
**Pull request by dblarons**
## Changes

- Bump to 3.0.0
- Remove `create` and `createTypes` because they are confusing and not that useful
- Rename `LazyTemplateProvider`->`LazyTemplateInterface` and replace the wacky getter/setter implementation with a single `tcomb.struct` that can be extended to add new factories more easily
- Define a method, `constructWithGetters` that takes a `tcomb.struct` and a value for that struct and creates getters for every prop
  - One advantage of the old `LazyTemplateProvider` was that fields that were removed without all references being removed would throw on missing getters when the module was loaded; that feature is retained here
  - Existing `setLazyTemplateFactory` calls **do not** need to be changed
- Update docs to reflect changes

### CODE REVIEW

File: src/__tests__/index.spec.js [hunk 1]
```diff
@@ -1,64 +1,42 @@
...
+    const builder = CharacterProfileBuilder.setLazyTemplateProvider(mockProvider);
+    expect(() => builder.getType()).to.not.throw();
+    expect(() => builder.getOptions()).to.not.throw();
   });
 
   describe('primitives', () => {
     it('are exported', () => {
-      expect(tcombBuilder.primitives.CheckboxBuilder).to.not.be.undefined;
-      expect(() => tcombBuilder.primitives.CheckboxBuilder.setLabel('foobar')).to.not.throw();
+      expect(primitives.CheckboxBuilder).to.not.be.undefined;
```
Comment (shawnthetaylor) [inline src/__tests__/index.spec.js:19]: Does undefined not have to be a function? I thought Enzyme switched everything to be a function call?

File: src/templates/__tests__/LazyTemplateInterface.spec.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      checkbox: MockTemplateFactory,
+      checkboxGroup: MockTemplateFactory,
+      dateTextField: MockTemplateFactory,
+      dropDown: MockTemplateFactory,
+      formPage: MockTemplateFactory,
+      radio: MockTemplateFactory,
+      staticPage: MockTemplateFactory,
+      struct: MockTemplateFactory,
+      textArea: MockTemplateFactory,
+      // textField: MockTemplateFactory, => missing
```
Comment (shawnthetaylor) [inline src/templates/__tests__/LazyTemplateInterface.spec.js:35]: delete

File: src/templates/__tests__/LazyTemplateInterface.spec.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      checkboxGroup: MockTemplateFactory,
+      dateTextField: MockTemplateFactory,
+      dropDown: MockTemplateFactory,
+      formPage: MockTemplateFactory,
+      radio: MockTemplateFactory,
+      staticPage: MockTemplateFactory,
+      struct: MockTemplateFactory,
+      textArea: MockTemplateFactory,
+      textField: MockTemplateFactory,
+
```
Comment (shawnthetaylor) [inline src/templates/__tests__/LazyTemplateInterface.spec.js:19]: extra space

File: src/templates/__tests__/LazyTemplateInterface.spec.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      extraField: MockTemplateFactory, // => extra
```
Comment (shawnthetaylor) [inline src/templates/__tests__/LazyTemplateInterface.spec.js:20]: delete comment?

File: src/templates/utils.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import tcomb from 'tcomb-validation';
+
+function toGetter(prop) {
+  const capsProp = prop.charAt(0).toUpperCase() + prop.slice(1);
```
Comment (shawnthetaylor) [inline src/templates/utils.js:4]: So annoying that Javascript doesn't have a titlecase method built-in... -_-

File: src/templates/utils.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+ * instantiate it. This is intended to be used when constructing the
+ * LazyTemplateInterface, where somewhat stronger typing can propogate errors
+ * earlier.
+ *
+ * @param {LazyTemplateInterface} TemplateInterface
+ * @param {object} instance
+ * @return {LazyTemplateInterface} A lazy template interface extended with
+ * getters.
+ */
+export function constructWithGetters(TemplateInterface, instance) {
```
Comment (liamcain) [inline src/templates/utils.js:20]: constructWithGetters is kind of a goofy name IMO but I can't think of anything better...

File: src/templates/utils.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  const interfaceProps = Object.keys(TemplateInterface.meta.props);
```
Comment (liamcain) [inline src/templates/utils.js:21]: What is `props` in this situation and how is it being used?

File: src/__tests__/index.spec.js (hunk 1 above)
Comment (dblarons) [reply src/__tests__/index.spec.js:19]: It's not unless `dirty-chai` is used, which is planned but out of scope for this PR.

https://github.com/cadre/tcomb-builder/issues/1

File: src/templates/__tests__/LazyTemplateInterface.spec.js (hunk 3 above)
Comment (dblarons) [reply src/templates/__tests__/LazyTemplateInterface.spec.js:19]: I put this space and the comment here to make it clearer to the reader that the extra field is extra. Is that not helpful to you as a reader?

File: src/templates/__tests__/LazyTemplateInterface.spec.js (hunk 2 above)
Comment (dblarons) [reply src/templates/__tests__/LazyTemplateInterface.spec.js:35]: I put this here to make it clearer that this is the missing field being tested. Is it not helpful for you?

File: src/templates/utils.js (hunk 7 above)
Comment (dblarons) [reply src/templates/utils.js:21]: They are a field on the `meta` object of an uninstantiated tcomb type: https://github.com/gcanti/tcomb/blob/master/docs/API.md#the-meta-object

That link talks about the meta object, and if you search for `props` you will see that `tcomb.interface` and `tcomb.struct` have `props` objects on their meta object.

File: src/templates/utils.js (hunk 6 above)
Comment (dblarons) [reply src/templates/utils.js:20]: agreed

File: src/templates/utils.js (hunk 5 above)
Comment (dblarons) [reply src/templates/utils.js:4]: agreed

**Review by liamcain — APPROVED** [review]

**Review by shawnthetaylor — APPROVED** [review]

**Pull request merged**